src/ir/subtype-exprs.h

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ir_subtype_exprs_h
+#define wasm_ir_subtype_exprs_h
+
+#include "ir/branch-utils.h"
+#include "wasm-traversal.h"
+#include "wasm.h"
+
+namespace wasm {
+
+//
+// Analyze subtyping relationships between expressions. This must CRTP with a
+// parent that implements:
+//
+//  * noteSubType(A, B) indicating A must be a subtype of B
+//  * noteCast(A, B) indicating A is cast to B
+//
+// There must be multiple versions of each of those, supporting A and B being
+// either a Type, which indicates a fixed type requirement, or an Expression*,
+// indicating a flexible requirement that depends on the type of that
+// expression. Specifically:
+//
+//  * noteSubType(Type, Type) - A constraint not involving expressions at all,
+//                              for example, an element segment's type must be
+//                              a subtype of the corresponding table's.
+//  * noteSubType(HeapType, HeapType) - Ditto, with heap types, for example in a
+//                                      CallIndirect.
+//  * noteSubType(Type, Expression) - A fixed type must be a subtype of an
+//                                    expression's type, for example, in BrOn
+//                                    (the declared sent type must be a subtype
+//                                    of the block we branch to).
+//  * noteSubType(Expression, Type) - An expression's type must be a subtype of
+//                                    a fixed type, for example, a Call operand
+//                                    must be a subtype of the signature's
+//                                    param.
+//  * noteSubType(Expression, Expression) - An expression's type must be a
+//                                          subtype of anothers, for example,
+//                                          a block and its last child.
+//
+//  * noteCast(HeapType, HeapType) - A fixed type is cast to another, for
+//  example,
+//                                   in a CallIndirect.
+//  * noteCast(Expression, Type) - An expression's type is cast to a fixed type,
+//                                 for example, in RefTest.
+//  * noteCast(Expression, Expression) - An expression's type is cast to
+//                                       another, for example, in RefCast.
```

Comment:
Perhaps I'll find out below, but what is the benefit of this over just extracting the type of the cast expression and using `noteCast(Expression, Type)` in this case?

Reply:
This PR tries to keep the info as "high-level" as possible, that is, not to lose information. We might want the extra information here to say something about the expression that fails to validate, if we use this in the validator some day (though, if we have the cast anyhow, we can do `cast->value`).